pkg/snapshot/copier/copier_test.go

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) 2018 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package copier_test
+
+import (
+	"os"
+	"reflect"
+
+	"github.com/gardener/etcd-backup-restore/pkg/miscellaneous"
+	. "github.com/gardener/etcd-backup-restore/pkg/snapshot/copier"
+	brtypes "github.com/gardener/etcd-backup-restore/pkg/types"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var (
+	copierConfig    *brtypes.CopierConfig
+	snapstoreConfig *brtypes.SnapstoreConfig
+	ss              brtypes.SnapStore
+	ds              brtypes.SnapStore
+	copier          *Copier
+)
+
+var _ = Describe("Running Copier", func() {
+	BeforeEach(func() {
+		copierConfig = NewCopierConfig()
+		copierConfig.SourceSnapstore = &brtypes.SnapstoreConfig{
+			MaxParallelChunkUploads: 5,
+			TempDir:                 "/tmp",
+			Provider:                "Local",
+			Container:               targetSnapstoreDir,
+		}
+		snapstoreConfig = &brtypes.SnapstoreConfig{
+			MaxParallelChunkUploads: 5,
+			TempDir:                 "/tmp",
+			Provider:                "Local",
+			Container:               snapstoreDir,
+		}
+
+		var err error
+		ds, ss, err = GetSourceAndDestinationStores(copierConfig, snapstoreConfig)
+		Expect(err).ToNot(HaveOccurred())
+		copier = NewCopier(ss, ds, logger)
+	})
+	It("should run the copy command succesfully", func() {
+		fullSourceStoreSnapshot, deltaSrourceStoreSnapList, err := miscellaneous.GetLatestFullSnapshotAndDeltaSnapList(ss)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(copier.Run()).ToNot(HaveOccurred())
+		fullTargetStoreSnapshot, deltaTargetStoreSnapList, err := miscellaneous.GetLatestFullSnapshotAndDeltaSnapList(ds)
+		Expect(snapsAreTheSame(fullSourceStoreSnapshot, fullTargetStoreSnapshot)).To(BeTrue())
+		for i, ssnap := range deltaSrourceStoreSnapList {
+			Expect(len(deltaSrourceStoreSnapList)).To(Equal(len(deltaTargetStoreSnapList)))
+			Expect(snapsAreTheSame(ssnap, deltaTargetStoreSnapList[i])).To(BeTrue())
+		}
+		_, err = os.Stat(targetSnapstoreDir)
+		Expect(err).ToNot(HaveOccurred())
+		os.Remove(targetSnapstoreDir)
+	})
+})
+
+func snapsAreTheSame(s1 *brtypes.Snapshot, s2 *brtypes.Snapshot) bool {
+	//SnapName and Prefix could be different after the copy operation.
+	logger.Logger.Infof("%v %v", s1, s2)
+
+	if !reflect.DeepEqual(s1.Kind, s2.Kind) {
```

Comment:
The problem with calling DeepEqual like this and just returning false is that we will not know what element caused the tests to fail. Perhaps you should change this method to instead call
`Expect(s1.Kind).To(Equal(s2.Kind))` and so on.

Reply:
Maybe [gomega gstruct](https://onsi.github.io/gomega/#gstruct-testing-complex-data-types)'s [MatchFields](https://pkg.go.dev/github.com/onsi/gomega/gstruct#MatchFields) is better suited in place of the `snapsAreTheSame` func.